### DISCUSSION TITLE
Proposal: Frontend plugins

### DISCUSSION THREAD

**Discussion by Haarolean** [kafbat/kafka-ui#232] [Tech Proposals]
| Column | Value |
|--|--|
| Status | Undecided |
| Impact | High |
| Driver | Need to provide API to include third party functionality to the app on frontend|
| Approver | - |
| Contributors | @Leshe4ka @Haarolean @germanosin |
| Started | 21.04.24 |
| Due date | - |
| Outcome | - |

## Background

**Context and Solution Overview**
In the realm of modern web development, embracing microfrontend architecture offers a granular approach to building scalable and maintainable applications. This proposal delves into the utilization of web components for constructing a microfrontend, specifically focusing on a remote application that integrates seamlessly with a host application.

**Development Requirements for the Remote Application**
The remote application is tasked with exporting an initialization function, which the host application will invoke. The core functionality of this function involves rendering application within a ShadowRoot container, ensuring style encapsulation and routing isolation.

**Initialization Function Overview (React example)**
Below is the initMicro script that outlines the initialization process:

```
import ReactDOM from "react-dom/client";
import React from "react";
import RemoteApp from "./RemoteApp";

const initMicro = (shadowRoot: ShadowRoot, routerBaseName: string) => {
  const container = shadowRoot.getElementById("root") || document.createElement("div");
  const styles = document.createElement("div");
  shadowRoot.appendChild(styles);
  const root = ReactDOM.createRoot(container);

  root.render(<RemoteApp styleContainer={styles} routerBaseName={routerBaseName} />);
};

export { initMicro };
```

The initMicro function accepts two parameters:

1. `shadowRoot` ([Shadow DOM MDN](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_shadow_DOM)): A container for the application's HTML content, ensuring style and script encapsulation.
2. `routerBaseName`: Specifies the base path for resolving routes within the microfrontend, critical for correct routing integration with the host application.

**Bundle Configuration and Asset Management**
To optimize loading and ensure correct asset serving, the remote application must configure its bundler to set a base path relative to its deployment location. For example, using Vite:

```
import { defineConfig } from "vite";

export default defineConfig({
  base: "./",
  build: {
    rollupOptions: {
      output: {
        chunkFileNames: (chunkInfo) => {
          if (chunkInfo.name === "initMicro") {
            return "initMicro.js";
          }

          return "[name]-[hash].js";
        },
      },
    },
  },
  
  ... // other config params
});
```

**Styling Considerations**
Styles within the remote application should reside inside the Shadow DOM to avoid conflicts and ensure correct application. Ideally, a single CSS file linked in the index.html suffices. If external libraries are used for styling (e.g., styled-components), they must support targeting elements within the Shadow DOM.

**Choosing External Libraries**
Selecting external libraries that interact with the DOM must be done with caution. These libraries should support rendering within the Shadow DOM or provide APIs to specify rendering containers (e.g., RadixUI). Libraries that default to manipulating the document's DOM directly could lead to functionality loss or style misapplication.

**Challenges and Potential Issues**

- _Shadow DOM Integration_: Ensuring all components and styles function correctly within a ShadowRoot can be challenging, particularly with third-party libraries not designed with Shadow DOM in mind.
- _Routing_: Coordinating routing between the host and remote applications requires careful implementation to ensure seamless navigation.
- _Performance_: Loading and initializing the remote application must be optimized to avoid performance bottlenecks, particularly in a microfrontend landscape where multiple applications coexist.

**Configuration File for Host Application Interaction**
To streamline the integration process between the host and remote applications, a configuration file, RemoteAppConfig, is essential. This configuration will guide the host application on how to locate, initialize, and render the remote applications within its environment.

**Configuration File Structure**
The configuration file is structured as follows:
```
interface RemoteAppConfig {
  apps: RemoteApp[];
}

interface RemoteApp {
  name: string; // The web component name, e.g., "remote-app"
  indexHTMLUrl: string; // Path to the index.html file, e.g., "https://your-site.com/index.html"
  initScriptUrl: string; // Path to the initialization script file, e.g., "https://your-site.com/initMicro.js"
  initScriptName: string; // Name of the exported initialization function, e.g., "initMicro"
  baseUrl: string; // e.g., "https://your-site.com/"
  route: RemoteRoute;
  tab: RemoteTab;
}

interface RemoteRoute {
  path: string; // Used in the UI routing to access the remote application, e.g., "your-site-ui"
}

interface RemoteTab {
  text: string; // Text displayed on the navigation button
  path: string; // Same as path in RemoteRoute
}
```

**Configuration Utilization**

- _Host Application Processing_: The host application uses this configuration to dynamically load and render remote applications. It fetches the index.html and the initialization script as defined, initializing each remote app within its designated container.
- _Dynamic Initialization_: By referencing initScriptUrl and initScriptName, the host application dynamically imports and invokes the initialization function, effectively bootstrapping the remote application.
- _Routing and Navigation_: The route and tab objects within the configuration enable the host application to integrate the remote application's routing seamlessly and provide user-friendly navigation options.

Benefits

- _Scalability_: Adding new remote applications or updating existing ones can be done by simply modifying the configuration file, without the need to alter the host application's codebase.
- _Decoupling_: This approach fosters a loosely coupled architecture, where the host and remote applications can evolve independently, enhancing maintainability.
- _Dynamic Integration_: The host application can conditionally load remote applications based on runtime criteria, optimizing resource usage and performance.

**Conclusion**
This proposal outlines a structured approach to integrating a remote application within a microfrontend architecture using web components. By adhering to the outlined practices, the development team can create isolated, maintainable, and scalable microfrontends that enhance the overall application architecture.

## Options considered 

| | Option 1: Web Components | Option 2: Framework-Specific Microfrontends | Option 3: Single Framework Microfrontends |
| - | - | - | - |
| Description | Utilize web components to build encapsulated microfrontends. | Leverage Module Federation for dynamic script loading and app integration at runtime. | Use iframes to isolate microfrontends within the host application. |
| Pros and Cons | _Pros_:<br>1. Framework-agnostic.<br>2. Encapsulates styles and behavior.<br>3. Reusable across different applications.<br>_Cons_:<br>1. Learning curve for new users.<br>2. Variability without strict conventions.<br>3. May require custom tooling for advanced features. | _Pros_:<br>1. Shared dependencies reduce load times.<br>2. Enables dynamic runtime integration.<br>3. Ideal for applications using the same framework.<br>_Cons_:<br>1. Tied to Webpack.<br>2. Complexity in setup and configuration.<br>3. Limited to JavaScript-based applications. | _Pros_:<br>1. Complete isolation ensures no style or script conflicts.<br>2. Simple to implement with basic HTML knowledge.<br>3. Framework and language agnostic.<br>_Cons_:<br>1. Overhead from multiple DOM trees.<br>2. Challenging to communicate between iframes.<br>3. Can affect the performance if not managed properly.

## Outcome

_Objective_: The proposal aims to integrate web components into the microfrontend architecture, leveraging their encapsulation and framework-agnostic features to enhance modularity, reusability, and maintainability of the application ecosystem.

**Key Outcomes**:
_Enhanced Modularity_:
Web components will provide a high degree of encapsulation, ensuring that individual microfrontends can be developed, updated, and maintained independently, reducing cross-team dependencies and streamlining the development process.

_Framework Agnosticism_:
The use of web components allows for flexibility in choosing or changing underlying frameworks or libraries across different microfrontends, facilitating a more adaptable and future-proof architecture.

_Scalability and Performance_:
By isolating microfrontends within web components, the proposal supports scalable and performant applications, enabling efficient loading and interaction without interference among components.

**Comment by siegenthalerroger** [comment]
Just wanted to chime in with a possible use-case for this, allowing custom functionality to be added to consumer management (in this example, "pausing" of consumers).

We have instrumented our (spring) kafka clients to have the capability of remotely being paused (or stopped). To achieve this, the clients consume a special internal topic, within which requested changes in the state of a consumer are stored. If we could have the capability of producing a message into this topic from a button in the UI (situated within the list of consumers, analagously to the offset management) this would be ideal.

**Comment by shlomiassaf** [comment]
This is a killer feature!

**Comment by marchenko1985** [comment]
I was wondering if there is a way to add some custom css

**use case:** read only mode, used by testers to see what messages are passing thru Kafka

**wish:** hide half of the functionality to make it simpler for testers to figure out how/where to navigate

**Comment by germanosin** [comment]
@mac2000, what specifically are you looking to hide?

**Reply by marchenko1985** [reply]
@germanosin imagine you are tester, and your are testing some new product feature, that involves kafka

if feature is not working, very first thing to check - was the message produced to the topic or not

thankfully this project has readonly mode, so you can be not afraid to click thru UI here and there in attempts to discover what's inside

but ideally, while we are in readonly mode, and for such use case - i probably would hide everything else, aka: all menus from sidebar except topics, all disabled buttons (while we are in readonly mode), all techy stuff from tables (like master node, in sync replicas etc)

so make it more friendly to non developer perspective

motivation: i believe developers, will use cli utils or admin client to get data they want, here, i just trying to make testers daily routine little bit easier - so we may say it is not about DX but more about TX 🙃

**Reply by germanosin** [reply]
@mac2000 
It seems like you could use RBAC to restrict access to all entities except topics and messages.
In this case, the UI will automatically hide the other menus.

> @germanosin imagine you are tester, and your are testing some new product feature, that involves kafka
> 
> if feature is not working, very first thing to check - was the message produced to the topic or not
> 
> thankfully this project has readonly mode, so you can be not afraid to click thru UI here and there in attempts to discover what's inside
> 
> but ideally, while we are in readonly mode, and for such use case - i probably would hide everything else, aka: all menus from sidebar except topics, all disabled buttons (while we are in readonly mode), all techy stuff from tables (like master node, in sync replicas etc)
> 
> so make it more friendly to non developer perspective
> 
> motivation: i believe developers, will use cli utils or admin client to get data they want, here, i just trying to make testers daily routine little bit easier - so we may say it is not about DX but more about TX 🙃
